artemis/src/main/java/tech/pegasys/artemis/cli/deposit/EncryptedKeystorePasswordProvider.java

```diff
@@ -0,0 +1,90 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.cli.deposit;
+
+import static org.apache.commons.lang3.StringUtils.isBlank;
+
+import com.google.common.io.Files;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import picocli.CommandLine.Model.CommandSpec;
+import picocli.CommandLine.ParameterException;
+
+public class EncryptedKeystorePasswordProvider {
+  private final String password;
+  private final String environmentVariable;
+  private final File passwordFile;
+  private final String errorPrefix;
+
+  private final CommandSpec spec;
+
+  public EncryptedKeystorePasswordProvider(
+      final CommandSpec spec,
+      final String password,
+      final String environmentVariable,
+      final File passwordFile,
+      final String errorPrefix) {
+    this.spec = spec;
+    this.password = password;
+    this.environmentVariable = environmentVariable;
+    this.passwordFile = passwordFile;
+    this.errorPrefix = errorPrefix;
+  }
+
+  public String retrievePassword() {
+    if (!isBlank(password)) {
+      return password;
+    }
+
+    if (environmentVariable != null) {
+      final String password = System.getenv(environmentVariable);
+      if (isBlank(password)) {
+        throw new ParameterException(
+            spec.commandLine(),
+            "Error in reading password from environment variable: " + environmentVariable);
+      }
+      return password;
+    }
+
+    if (passwordFile != null) {
+      try {
+        final String password =
+            Files.asCharSource(passwordFile, StandardCharsets.UTF_8).readFirstLine();
+        if (isBlank(password)) {
+          throw new ParameterException(
+              spec.commandLine(), "Empty password read from password file: " + passwordFile);
+        }
+        return password;
+      } catch (final FileNotFoundException e) {
+        throw new ParameterException(
+            spec.commandLine(), "Password file not found: " + passwordFile);
+      } catch (IOException e) {
+        throw new ParameterException(
+            spec.commandLine(),
+            "Unexpected IO error while reading password from file ["
+                + passwordFile
+                + "] : "
+                + e.getMessage());
+      }
+    }
+
+    final String errorMessage =
+        String.format(
+            "Missing options: [--%1$s-password | --%1$s-password:env | --%1$s-password:file]",
+            errorPrefix);
+    throw new ParameterException(spec.commandLine(), errorMessage);
```

Comment:
It's really annoying to get the missing options error message in small chunks.  ie I tried:
```
$ ./build/install/teku/bin/teku validator generate
Missing required options [--node-url=<URL>, --contract-address=<ADDRESS>, --private-key=<KEY>]

$ ./build/install/teku/bin/teku validator generate --node-url=http://localhost:8545 --contract-address=dddddddd --private-key=222222222
Missing options: [--validator-password | --validator-password:env | --validator-password:file]

$ ./build/install/teku/bin/teku validator generate --node-url=http://localhost:8545 --contract-address=dddddddd --private-key=222222222 --validator-password
Enter value for --validator-password (Password to encrypt validator keys):
Missing options: [--withdrawal-password | --withdrawal-password:env | --withdrawal-password:file]
```
So it took me four goes to actually be told all of what I had to provide.

Reply:
I have introduced mutually exclusive ArgGroup so that picocli handles these errors automagically. I do have to extract out subcommands by converting methods to standalone classes to make ArgGroup work (I believe PicoCli has a bug where ArgGroup doesn't work if methods are used as subcommands). I have also changed the `--encrypt-keys` to `--encrypted-keystore-enabled` and we now run in interactive mode to obtain passwords if none of the non-interactive password options are specified. The ArgGroup also allows to provide sub-section heading. e.g. 

~~~
./teku validator generate --help

teku validator generate [OPTIONS]

Description:

Register validators by generating new keys and sending deposit transactions to
an Ethereum 1 node

Options:
  -a, --amount=<GWEI>       Deposit amount in Gwei (default: 32000000000)
  -c, --contract-address=<ADDRESS>
                            Address of the deposit contract
  -e, --encrypted-keystore-enabled=<true|false>
                            Create encrypted keystores for validator and
                              withdrawal keys. (Default: true)
  -h, --help                Show this help message and exit.
  -n, --number-of-validators=<NUMBER>
                            The number of validators to create keys for and
                              register
  -o, --output-path=<FILE|DIR>
                            Path to output file for unencrypted keys or output
                              directory for encrypted keystore files. If not
                              set, unencrypted keys will be written on standard
                              out and encrypted keystores will be created in
                              current directory
  -p, --private-key=<KEY>   Ethereum 1 private key to use to send transactions
  -u, --node-url=<URL>      JSON-RPC endpoint URL for the Ethereum 1 node to
                              send transactions via
  -V, --version             Print version information and exit.
Non-interactive password options for validator keystores:
      --validator-password-env=<ENV_VAR>
                            Read password from environment variable to encrypt
                              the validator keys
      --validator-password-file=<FILE>
                            Read password from the file to encrypt the
                              validator keys
Non-interactive password options for withdrawal keystores:
      --withdrawal-password-env=<ENV_VAR>
                            Read password from environment variable to encrypt
                              the withdrawal keys
      --withdrawal-password-file=<FILE>
                            Read password from the file to encrypt the
                              withdrawal keys

Teku is licensed under the Apache License 2.0
~~~

(@jakehaugen and @arash009  FYI)

Reply:
Running generate command with required parameters only:

~~~
./teku validator generate --contract-address=dddddddddddddddddddddddddddddddddddddddd \
--number-of-validators=1 \
--private-key=8f2a55949038a9610f50fb23b5883af3b4ecb3c3bb792cbcefbd1542c692be63 \
--node-url=http://localhost:8545
Enter password for Validator Keystore:
Re-Enter password for Validator Keystore:
Enter password for Withdrawal Keystore:
Re-Enter password for Withdrawal Keystore:
22:18:24.268 [INFO ] - Generating Encrypted Keystores in .
~~~

Reply:
To confirm, do you have to use the same method to supply the validator password and withdrawal password? e..g. can I supply validator password via env variable, and use interactive method for withdrawal password? (I dont think there really is a use case for that, but just checking what works)

Reply:
@arash009  yes, both validator keystore and withdrawal keystore options are independent of each other, so validator keystore's password can come from environment variable for instance, while withdrawal's password can be read from file. However, due to mutually exclusive mode, you can either specify `--validator-password-env` or `--validator-password-file` but not both.